### DISCUSSION TITLE
Electron-Builder installers for multi platforms

### DISCUSSION THREAD

**Discussion by mandolyte** [community/community#25177] [Actions]
<p>Building an electron app on three platforms. The tool <code>electron-builder</code> will produce installers with names like these two for Ubuntu and Windows:</p>
<ul>
<li>
<p><code>my-app\_1.0.1\_amd64.deb</code></p>
</li>
<li>
<p><code>my-app Setup 1.0.1.exe</code></p>
</li>
</ul>
<p>The version string comes its <code>package.json</code>. And I want to run the workflow when I push the tag for <code>v1.0.1</code>.</p>
<p>I have not been able to figure out a way to set the artifact and asset names to contain the version string (not semver).</p>
<p>At the moment, I am hard-coding the version into the workflow itself. Then using action <code>svenstaro/upload-release-action@v1-release</code> to upload the tag.</p>
<p>Here is the start of the workflow and below that is the uploading action.</p>
<pre><code class="lang-auto">
name:&amp;nbsp;build and release electron from react web  

on:
&amp;nbsp;&amp;nbsp;push:
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;tags:
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;-&amp;nbsp;'v\*'&amp;nbsp;#&amp;nbsp;any push with a&amp;nbsp;tag in semver format
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;
jobs:
&amp;nbsp;&amp;nbsp;build:
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;name:&amp;nbsp;build-electron-from-react-web
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;env:
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;VERSION:&amp;nbsp;1.0.1  

&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;#&amp;nbsp;update versions below
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;runs-on:&amp;nbsp;${{&amp;nbsp;matrix.os&amp;nbsp;}}
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;strategy:
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;matrix:
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;os:&amp;nbsp;[ubuntu-latest,&amp;nbsp;windows-latest,&amp;nbsp;macos-latest]
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;include:
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;-&amp;nbsp;os:&amp;nbsp;ubuntu-latest
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;artifact\_name:&amp;nbsp;book-package-app\_1.0.1\_amd64.deb
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;asset\_name:&amp;nbsp;book-package-app\_1.0.1\_amd64.deb
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;-&amp;nbsp;os:&amp;nbsp;windows-latest
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;artifact\_name:&amp;nbsp;"book-package-app Setup 1.0.1.exe"
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;asset\_name:&amp;nbsp;"book-package-app Setup 1.0.1.exe"
</code></pre>
<p>The uploading action:</p>
<pre><code class="lang-auto">
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;-&amp;nbsp;name:&amp;nbsp;Upload Release Assets
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;uses:&amp;nbsp;svenstaro/upload-release-action@v1-release
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;env:
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;GITHUB\_TOKEN:&amp;nbsp;${{&amp;nbsp;secrets.GITHUB\_TOKEN&amp;nbsp;}}
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;with:
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;repo\_token:&amp;nbsp;${{&amp;nbsp;secrets.GITHUB\_TOKEN&amp;nbsp;}}
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;file:&amp;nbsp;"./electron/dist/${{&amp;nbsp;matrix.artifact\_name&amp;nbsp;}}"
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;asset\_name:&amp;nbsp;${{&amp;nbsp;matrix.asset\_name&amp;nbsp;}}
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;tag:&amp;nbsp;${{&amp;nbsp;github.ref&amp;nbsp;}}
&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;overwrite:&amp;nbsp;true
</code></pre>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi <a class="mention" href="/mandolyte">@mandolyte</a> ,</p>
<p>Thank you for being here!</p>
<ol>
<li>‘Workflow&amp;Job level env’ and matrix doesn’t accept variables, hence you can set the artifacts/asset name in step level.</li>
</ol>
<p>2. The version string comes its <code>package.json</code>. And I want to run the workflow when I push the tag for <code>v1.0.1</code>.</p>
<p>&gt;&gt; You can get the tag version from ‘${GITHUB_REF##*v}’, and compare the value in ‘package.json’, use if expression to execute the next steps accordingly.</p>
<p>In total, i assume your script can be changed as below(I removed macos part you can add accordingly).</p>
<pre><code>jobs:
  build:
    name: build-electron-from-react-web
# env:
# VERSION: 1.0.1

    # update versions below
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        os: [ubuntu-latest, windows-latest]
 # include:
 # - os: ubuntu-latest
 # artifact_name: book-package-app_1.0.1_amd64.deb
 # asset_name: book-package-app_1.0.1_amd64.deb
 # - os: windows-latest
 # artifact_name: "book-package-app Setup 1.0.1.exe"
 # asset_name: "book-package-app Setup 1.0.1.exe"

    steps:
      - name: checkout
        uses: actions/checkout@v2
      - name: set tag version #If needed, here you can can compare the value form package.json with ${GITHUB_REF##*v}.
        shell: bash
        run: |
          echo "::set-env name=version::${GITHUB_REF##*v}"

      - name: set artifact&amp;asset name for ubuntu
        if: runner.os=='linux'
        run: |
          echo "::set-env name=artifactname::book-package-app_${version}_amd64.deb"
          echo "::set-env name=assetname::book-package-app_${version}_amd64.deb"
      - name: set artifact&amp;asset name for windows
        shell: bash
        if: runner.os=='windows'
        run: |
          echo "::set-env name=artifactname::book-package-app Setup ${version}.exe"
          echo "::set-env name=assetname::book-package-app Setup ${version}.exe"
      - name: Upload Release Assets
        uses: svenstaro/upload-release-action@v1-release
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        with:
          repo_token: ${{ secrets.GITHUB_TOKEN }}
          file: "./electron/dist/${{ env.artifactname }}"
          asset_name: ${{ env.assetname }}
          tag: ${{ github.ref }}
          overwrite: true
</code></pre>
<p>Hope it helps!</p>

**Comment by mandolyte** [comment]
<p><a class="mention" href="/weide-zhou">@weide-zhou</a> </p>
<p>I think I spoke too soon. Looks like env vars are not allowed where I need them. Hope you have another idea!</p>
<p><img src="https://user-images.githubusercontent.com/1289437/181089253-b09c207f-7ff0-4eb2-bc31-2621ca8ca66c.png" alt="Annotation 2020-03-26 112754.png" data-base62-sha1="qfrSksRfYU0zTEl1yAIVsd9AhIx" width="614" height="278"></p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/mandolyte">@mandolyte</a> ,</p>
<p>Thanks for your reply! The env value to get in action ‘with’ should use the format : ${{ env.artifactname }}</p>
<p>I have changed in my first post, thanks.</p>
<p><img src="https://user-images.githubusercontent.com/52772761/181089256-bf799fb2-eefe-4d18-86d0-66130b5373b0.png" alt="envvalue.png" data-base62-sha1="jF8XCPV56YCZPIzVYEewtzYbghM" width="482" height="441"></p>
